### PR TITLE
Allow shorter signed hex literals with appropriate type suffixes

### LINKED ISSUE #11823: Negative hex literal int16 throws parse error
# Steps to reproduce

```powershell
0xACEDs
```

# Expected behavior

```none
-21267
```

# Actual behavior

```none
ParserError:
Line |
   1 |  0xACEDs
     |         ~
     | The numeric constant 0xACEDs is not valid.
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.0-rc.2
PSEdition                      Core
GitCommitId                    7.0.0-rc.2
OS                             Microsoft Windows 10.0.18362
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by vexx32**
# PR Summary

This change allows shorter hex literals to be treated as signed, if the specified type suffix is a signed type with correct hex lengths. For example, `0xFFFFs` now parses correctly to -1 instead of erroring. Default behaviour without a specific suffix and/or with explicit unsigned suffixes remains as it was. 

## PR Context

Resolves #11823. See the issue for further context and discussion.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 1]
```diff
@@ -1460,17 +1460,28 @@ private char Backtick(char c, out char surrogateCharacter)
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: @vexx32, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Basic Auth over HTTP not allowed on Unix.New-PSSession should NOT throw a ConnectFailed exception when specifying Basic Auth over HTTPS on Unix
```powershell
Expected 2148734209, but got -2146233087.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1: line 61
61:         $err.Exception.HResult | Should -Be 0x80131501
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0x80000000 should return -2147483648, with type [int]
```powershell
Expected -2147483648, but got 2147483648.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFF should return -1, with type [int]
```powershell
Expected '-1', but got 4294967295.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFFFFFFFFFF should return -1, with type [long]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:19
+ 0xFFFFFFFFFFFFFFFF
+                   ~
The numeric constant 0xFFFFFFFFFFFFFFFF is not valid."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFy should return -1, with type [sbyte]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:6
+ 0xFFy
+      ~
The numeric constant 0xFFy is not valid."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```

Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: @vexx32, your last commit had 7 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Basic Auth over HTTP not allowed on Unix.New-PSSession should NOT throw a ConnectFailed exception when specifying Basic Auth over HTTPS on Unix
```powershell
Expected 2148734209, but got -2146233087.
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1: line 61
61:         $err.Exception.HResult | Should -Be 0x80131501
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0x80000000 should return -2147483648, with type [int]
```powershell
Expected -2147483648, but got 2147483648.
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFF should return -1, with type [int]
```powershell
Expected '-1', but got 4294967295.
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFFFFFFFFFF should return -1, with type [long]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:19
+ 0xFFFFFFFFFFFFFFFF
+                   ~
The numeric constant 0xFFFFFFFFFFFFFFFF is not valid."
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFy should return -1, with type [sbyte]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:6
+ 0xFFy
+      ~
The numeric constant 0xFFy is not valid."
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```

Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: @vexx32, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0x80000000 should return -2147483648, with type [int]
```powershell
Expected -2147483648, but got 2147483648.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFF should return -1, with type [int]
```powershell
Expected '-1', but got 4294967295.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFFFFFFFFFF should return -1, with type [long]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:19
+ 0xFFFFFFFFFFFFFFFF
+                   ~
The numeric constant 0xFFFFFFFFFFFFFFFF is not valid."
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFy should return -1, with type [sbyte]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:6
+ 0xFFy
+      ~
The numeric constant 0xFFy is not valid."
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFs should return -1, with type [short]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:8
+ 0xFFFFs
+        ~
The numeric constant 0xFFFFs is not valid."
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 2]
```diff
@@ -3685,9 +3697,41 @@ private void ScanNumberAfterDot(StringBuilder sb, ref int signIndex, ref bool no
...
+
+                                if (strNum.Length == expectedLength)
+                                {
+                                    strNum = strNum.Slice(1);
+                                }
+                            }
+
+
+                            // If we're expecting a sign bit, remove the leading 0 added in ScanNumberHelper
+                            if (!suffix.HasFlag(NumberSuffixFlags.Unsigned))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: Why is this a dup of 3700 line?

Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Oh, _weird_. Guessing a copy-paste error. Thanks for noticing that! I'll edit the duplicate segment out. 💖

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 1 above)
Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: @vexx32, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Basic Auth over HTTP not allowed on Unix.New-PSSession should NOT throw a ConnectFailed exception when specifying Basic Auth over HTTPS on Unix
```powershell
Expected 2148734209, but got -2146233087.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1: line 61
61:         $err.Exception.HResult | Should -Be 0x80131501
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0x80000000 should return -2147483648, with type [int]
```powershell
Expected -2147483648, but got 2147483648.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFF should return -1, with type [int]
```powershell
Expected '-1', but got 4294967295.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFFFFFFFFFF should return -1, with type [long]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:19
+ 0xFFFFFFFFFFFFFFFF
+                   ~
The numeric constant 0xFFFFFFFFFFFFFFFF is not valid."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFy should return -1, with type [sbyte]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:6
+ 0xFFy
+      ~
The numeric constant 0xFFy is not valid."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```

Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: @vexx32, your last commit had 7 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Basic Auth over HTTP not allowed on Unix.New-PSSession should NOT throw a ConnectFailed exception when specifying Basic Auth over HTTPS on Unix
```powershell
Expected 2148734209, but got -2146233087.
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1: line 61
61:         $err.Exception.HResult | Should -Be 0x80131501
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0x80000000 should return -2147483648, with type [int]
```powershell
Expected -2147483648, but got 2147483648.
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFF should return -1, with type [int]
```powershell
Expected '-1', but got 4294967295.
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFFFFFFFFFF should return -1, with type [long]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:19
+ 0xFFFFFFFFFFFFFFFF
+                   ~
The numeric constant 0xFFFFFFFFFFFFFFFF is not valid."
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFy should return -1, with type [sbyte]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:6
+ 0xFFy
+      ~
The numeric constant 0xFFy is not valid."
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```

Comment (PoshChan) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: @vexx32, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0x80000000 should return -2147483648, with type [int]
```powershell
Expected -2147483648, but got 2147483648.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFF should return -1, with type [int]
```powershell
Expected '-1', but got 4294967295.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 1143
1143:             ExecuteCommand $Script | Should -Be $ExpectedValue
```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFFFFFFFFFFFFF should return -1, with type [long]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:19
+ 0xFFFFFFFFFFFFFFFF
+                   ~
The numeric constant 0xFFFFFFFFFFFFFFFF is not valid."
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFy should return -1, with type [sbyte]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:6
+ 0xFFy
+      ~
The numeric constant 0xFFy is not valid."
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Numerical Notations Tests (starting at line 2374 to line 2452).0xFFFFs should return -1, with type [short]
```powershell
Exception calling "Invoke" with "0" argument(s): "At line:1 char:8
+ 0xFFFFs
+        ~
The numeric constant 0xFFFFs is not valid."
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Parser.Tests.ps1: line 81
81:                 $PowerShell.AddScript($command).Invoke()

```

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 3]
```diff
@@ -4988,7 +5012,8 @@ internal Token NextToken()
                         _currentIndex = _tokenStart;
                         c = GetChar();
 
-                        if (strNum == null) { return ScanGenericToken(c); }
+                        if (strNum == null)
+                        { return ScanGenericToken(c); }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: Formatting.

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

**Comment by adityapatwardhan** [comment]
@daxian-dbw The change is in the tokenizer, I would like you to have a look.

**Review by adityapatwardhan — APPROVED** [review]

**Review by daxian-dbw — APPROVED** [review]

**Comment by iSazonov** [comment]
BackPort-7.0.x-Consider is asked in #14302

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT / @joeyaiello We need you comment about whether this should be back ported to 7.0.x?

**Comment by TravisEz13** [comment]
@SteveL-MSFT Do we want to backport this?

**Pull request merged**